src/ORM/DataList.php

```diff
@@ -1144,139 +1141,157 @@ private function fetchEagerLoadRelations(Query $query): void
         }
+
         return [$relationArray, $relationItemIDs];
     }
 
     private function fetchEagerLoadHasMany(
-        array $ids,
+        array $parentIDs,
         string $hasManyIDField,
         string $relationDataClass,
         string $eagerLoadRelation,
-        string $relation
-    ): array
-    {
-        $relationArray = DataObject::get($relationDataClass)->filter([$hasManyIDField => $ids])->toArray();
+        string $relationName
+    ): array {
+        // Get ALL of the items for this relation up front, for ALL of the parents
+        // Fetched as an array to avoid sporadic additional queries when the DataList is looped directly
+        $relationArray = DataObject::get($relationDataClass)->filter([$hasManyIDField => $parentIDs])->toArray();
         $relationItemIDs = [];
+
+        // Store the children in an ArrayList against the correct parent
         foreach ($relationArray as $relationItem) {
             $relationItemIDs[] = $relationItem->ID;
             $eagerLoadID = $relationItem->$hasManyIDField;
-            if (!isset($this->eagerLoadedData[$eagerLoadRelation][$eagerLoadID][$relation])) {
+            if (!isset($this->eagerLoadedData[$eagerLoadRelation][$eagerLoadID][$relationName])) {
                 $arrayList = ArrayList::create();
                 $arrayList->setDataClass($relationItem->dataClass);
-                $this->eagerLoadedData[$eagerLoadRelation][$eagerLoadID][$relation] = $arrayList;
+                $this->eagerLoadedData[$eagerLoadRelation][$eagerLoadID][$relationName] = $arrayList;
             }
-            $this->eagerLoadedData[$eagerLoadRelation][$eagerLoadID][$relation]->push($relationItem);
+            $this->eagerLoadedData[$eagerLoadRelation][$eagerLoadID][$relationName]->push($relationItem);
         }
+
         return [$relationArray, $relationItemIDs];
     }
 
     private function fetchEagerLoadManyMany(
         array $manyManyLastComponent,
-        array $ids,
+        array $parentIDs,
         string $relationDataClass,
         string $eagerLoadRelation,
-        string $relation
-    ): array
-    {
-        $parentField = $manyManyLastComponent['parentField'];
-        $childField = $manyManyLastComponent['childField'];
+        string $relationName,
+        string $parentDataClass
+    ): array {
+        $parentIDField = $manyManyLastComponent['parentField'];
+        $childIDField = $manyManyLastComponent['childField'];
         // $join will either be:
         // - the join table name for many-many
         // - the join data class for many-many-through
         $join = $manyManyLastComponent['join'];
+
         // many_many_through
         if (is_a($manyManyLastComponent['relationClass'], ManyManyThroughList::class, true)) {
-            $joinThroughObjs = $join::get()->filter([$parentField => $ids]);
+            $joinThroughObjs = DataObject::get($join)->filter([$parentIDField => $parentIDs]);
```

Comment:
Changed to be more consistent with the other `DataObject::get()` calls in the eager loading code